README.md
Description
===========

This cookbook provides two resources and providers to install and configure [Graphite](http://graphite.wikidot.com/) *Carbon {Cache:Relay}* under [virtualenv](http://pypi.python.org/pypi/virtualenv).  Currently supported resources:

* Carbon Relay (`relay`)
* Carbon Cache (`cache`)

Requirements
============

1. [Upstart](http://upstart.ubuntu.com/)
  Pull requests accepted to support other init styles
2. [Python](http://github.com/opscode-cookbooks/python/)
  Provides virtualenv support and the pip provider
3. [graphite](http://github.com/damm/graphite/)
  Installs the **Graphite Web** Service to display Metrics stored by *Carbon Cache*

Recipes
============

default.rb
----------

The default recipe does absolutely nothing.  It is not intended to do anything 

Resources and Providers
=======================

This cookbook provides two resource and corresponding providers.

`cache.rb`
-------------

Installs and Configures [Carbon Cache](https://github.com/graphite-project/carbon) from [Pypi](http://pypi.python.org/pypi/carbon)

Actions:

* `install` - installs carbon-cache
* `config` - configures carbon-cache
* `start` - starts carbon-cache
* `stop` - stops carbon-cache

Attribute Parameters:

* `cookbook` - String - default - `carbon` - cookbook to fetch *all* templates from
* `python_interpreter` - String - default - `python2.7`
* `carbon_template_source` - String - default - `carbon.conf.erb`
* `storage_template_source` - String -  default - `storage-schemas.conf.erb`
* `cpu_affinity` - String
* `options` - Hash
* `user` - String - default - `graphite`
* `group` - String - default - `graphite`
* `max_cache_size` - String - default - `inf`
* `max_updates_per_second` - Fixnum - default - `5000`
* `max_creates_per_minute` - Fixnum - default - `100`
* `line_listner` - Hash - default - `{"line_receiver_interface" => "0.0.0.0", "line_receiver_port" => 2003 }`
* `pickle_listner` - Hash - default - `{"pickle_receiver_interface" => "0.0.0.0", "pickle_receiver_port" => 2004 }`
* `udp_listner` - Hash - default - `{"enable_udp_listner" => "False", "udp_receiver_interface" => "0.0.0.0", "udp_receiver_port" => 2003 }`
* `use_insecure_pickler` - String - default - `False`
* `use_flow_control` - String - default - `True`
* `cache_query` - Hash -  default - `{"cache_query_interface" => "0.0.0.0", "cache_query_port" => 7002 }`
* `log_updates` - String - default - `False`
* `whisper_autoflush` - String -  default - `False`
* `enable_amqp` - String -  default - `False`
* `amqp_verbose` - String -  default - `False`
* `amqp` - Hash - default - `{"amqp_host" => "localhost", "amqp_port" => 5672, "amqp_vhost" => "/", "amqp_user" => "guest", "amqp_password" => "guest", "amqp_exchange" => "graphite", "amqp_metric_name_in_body" => "False" }`
* `bind_patterns` - Array -  default - `"#"`
* `graphite_home` - String - default - `/opt/graphite`
* `carbon_packages` - Hash -  default - `{ "whisper" => "0.9.10", "carbon" => "0.9.10" }`
* `carbon_instance` - String -  default - `a`
* `storage_schema` - Hash - default - `{ :all => { :pattern => "(.*)", :retentions => "10s:90d, 60s:1y" } }`
* `local_data_dir` - String - default - `/opt/graphite/storage/whisper`
* `init_style` - String - default - `upstart`

`relay.rb`
-------------

Configures [Carbon Relay](https://github.com/graphite-project/carbon)


Actions:

* `config` - configures carbon-relay
* `start` - starts carbon-relay
* `stop` - stops carbon-relay
* `disable` - disables carbon-relay

Attribute Parameters:

* `cookbook` -  String - `carbon` - name of the cookbook to get templates from
* `options` -  Hash 
* `relay_rules` -  Hash - default - `{"default" => { "default" => "true"' - "destinations" => ["127.0.0.1:2004"]' - "continue" => "false"' - "pattern" => String.new } }`
* `line_listner` -  Hash - default - `{"line_receiver_interface" => "0.0.0.0"' - "line_receiver_port" => 2013 }`
* `pickle_listner` -  Hash - default `{"pickle_receiver_interface" => "0.0.0.0"' - "pickle_receiver_port" => 2014 }`
* `relay_method` -  String - default - `consistent-hashing` 
* `replication_factor` -  Fixnum - default - `1`
* `destinations` -  Array' - default - `["127.0.0.1:2004:a"]`
* `max_datapoints_per_message` -  Fixnum - default `500`
* `max_queue_size` -  Fixnum - default - `10000`
* `use_flow_control` -  String - default - `True`
* `graphite_home` -  String - default - `/opt/graphite` - location to install graphite to
* `relay_instance` -  String - default - `a` 
* `user` - String - default - `graphite` - username to install graphite under
* `group` - String - default -  `graphite` - group to install graphite under
* `python_interpreter` -  String - default - `python2.7`
* `relay_template_source` -  String - default - `carbon-relay.conf.erb`
* `relay_init_template` -  String - default - `carbon-relay.init.erb`
* `init_style` -  String' - default - `upstart` 
* `cpu_affinity` -  String - Specifies which cpu to bind the carbon-relay process to
* `relay_rules_template_source` -  String - default - `relay-rules.conf.erb`

Usage
==================

This is the most basic example you are not limited by this example.

```cd chef-repo/cookbooks```  
```git submodule add git://github.com/damm/carbon.git```  
```git submodule add git://github.com/damm/graphite.git```  
```COOKBOOK=graphite_infra rake new_cookbook```  

* The cookbook named ``graphite_infra`` must depend on the *carbon* cookbook.

    carbon_cache "my_carbon" do
      action [:install,:config,:start]  
      cpu_affinity "1"  
    end  
    carbon_relay "sample" do  
      action [:config,:start]    
    end


License and Author
==================

Author:: Scott M. Likens <scott@spam.likens.us>


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
  


CHANGELOG.md
# CHANGELOG for carbon

This file is used to list changes made in each version of carbon.

## 0.1.0:

* Work in Progress version of carbon


resources/relay.rb
actions :config, :start, :stop, :disable

attribute :cookbook, :kind_of => String, :default => "carbon"
attribute :options, :kind_of => Hash
attribute :relay_rules, :kind_of => Hash, :default => { "default" => { "default" => "true", "destinations" => ["127.0.0.1:2004"], "continue" => "false", "pattern" => String.new } }
attribute :line_listner, :kind_of => Hash, :default => {"line_receiver_interface" => "0.0.0.0", "line_receiver_port" => 2013 }
attribute :pickle_listner, :kind_of => Hash, :default => {"pickle_receiver_interface" => "0.0.0.0", "pickle_receiver_port" => 2014 }
attribute :relay_method, :kind_of => String, :default => "consistent-hashing"
attribute :replication_factor, :kind_of => Fixnum, :default => 1
attribute :destinations, :kind_of => Array, :default => ["127.0.0.1:2004:a"]
attribute :max_datapoints_per_message, :kind_of => Fixnum, :default => 500
attribute :max_queue_size, :kind_of => Fixnum, :default => 10000
attribute :use_flow_control, :kind_of => String, :default => "True"
attribute :graphite_home, :kind_of => String, :default => "/opt/graphite"
attribute :relay_instance, :kind_of => String, :default => "a"
attribute :user, :kind_of => String, :default => "graphite"
attribute :group, :kind_of => String, :default => "graphite"
attribute :python_interpreter, :kind_of => String, :default => "python2.7"
attribute :relay_template_source, :kind_of => String, :default => "carbon-relay.conf.erb"
attribute :relay_init_template, :kind_of => String, :default => "carbon-relay.init.erb"
attribute :init_style, :kind_of => String, :default => "upstart"
attribute :cpu_affinity, :kind_of => String
attribute :relay_rules_template_source, :kind_of => String, :default => "relay-rules.conf.erb"

def initialize(*args)
  super
  @action = :nothing
end


resources/cache.rb
actions :install, :config, :start, :stop

# No Dependency checking here, this obviously will not be true on distros that do not run 2.7
attribute :cookbook, :kind_of => String, :default => "carbon"
attribute :python_interpreter, :kind_of => String, :default => "python2.7"
attribute :carbon_template_source, :kind_of => String, :default => "carbon.conf.erb"
attribute :storage_template_source, :kind_of => String, :default => "storage-schemas.conf.erb"
attribute :cpu_affinity, :kind_of => String
attribute :options, :kind_of => Hash
attribute :user, :kind_of => String, :default => "graphite"
attribute :group, :kind_of => String, :default => "graphite"
attribute :max_cache_size, :kind_of => String, :default => "inf"
attribute :max_updates_per_second, :kind_of => Fixnum, :default => 5000
attribute :max_creates_per_minute, :kind_of => Fixnum, :default => 100
attribute :line_listner, :kind_of => Hash, :default => {"line_receiver_interface" => "0.0.0.0", "line_receiver_port" => 2003 }
attribute :pickle_listner, :kind_of => Hash, :default => {"pickle_receiver_interface" => "0.0.0.0", "pickle_receiver_port" => 2004 }
attribute :udp_listner, :kind_of => Hash, :default => {"enable_udp_listner" => "False", "udp_receiver_interface" => "0.0.0.0", "udp_receiver_port" => 2003 }
attribute :use_insecure_pickler, :kind_of => String, :default => "False"
attribute :use_flow_control, :kind_of => String, :default => "True"
attribute :cache_query, :kind_of => Hash, :default => {"cache_query_interface" => "0.0.0.0", "cache_query_port" => 7002 }
attribute :log_updates, :kind_of => String, :default => "False"
attribute :whisper_autoflush, :kind_of => String, :default => "False"
attribute :enable_amqp, :kind_of => String, :default => "False"
attribute :amqp_verbose, :kind_of => String, :default => "False"
attribute :amqp, :kind_of => Hash, :default => {"amqp_host" => "localhost", "amqp_port" => 5672, "amqp_vhost" => "/", "amqp_user" => "guest", "amqp_password" => "guest", "amqp_exchange" => "graphite", "amqp_metric_name_in_body" => "False" }
attribute :bind_patterns, :kind_of => Array, :default => ["#"]
attribute :graphite_home, :kind_of => String, :default => "/opt/graphite"
attribute :carbon_packages, :kind_of => Hash, :default => { "whisper" => "0.9.10", "carbon" => "0.9.10" }
attribute :carbon_instance, :kind_of => String, :default => "a"
attribute :storage_schema, :kind_of => Hash, :default => { :all => { :pattern => "(.*)", :retentions => "10s:90d, 60s:1y" } }
attribute :local_data_dir, :kind_of => String, :default => "/opt/graphite/storage/whisper"
attribute :init_style, :kind_of => String, :default => "upstart"

def initialize(*args)
  super
  @action = :install
  @run_context.include_recipe ["build-essential","python","python::pip","python::virtualenv"]
end
